LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceCache.cs

```diff
@@ -0,0 +1,344 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaTools.LoRaMessage;
+    using LoRaTools.Utils;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public class LoRaDeviceCache : IDisposable
+    {
+        private readonly LoRaDeviceCacheOptions options;
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, LoRaDevice>> devAddrCache = new();
```

Comment:
Out of interrest, as we are always locking, what is the advantage of using a concurrent dict here?

Reply:
We do allow concurrent access when refreshing the cache. Hence, we need it there.